### PR TITLE
Upgrade imgui to v1.82

### PR DESCRIPTION
**Pull request by ptxmac**
This upgrades to imgui v1.82. I didn't do extensive testing, but some of my own projects worked without any issue.

The only weirdness was the need to disable some default "ime" windows handlers (see wrapper.go). I don't know if that's only a problem with github ci, or also on a real window machine.

In the v1.81 the "ime" stuff was guarded by the `__GNUC__` flag, so it was also disabled previously

I also added go1.16 to the test matrix

### CODE REVIEW

File: imgui/imgui.cpp [hunk 1]
```diff
@@ -10620,7 +10690,7 @@ static void SetClipboardTextFn_DefaultImpl(void*, const char* text)
 #endif
 
 // Win32 API IME support (for Asian languages, etc.)
-#if defined(_WIN32) && !defined(__GNUC__) && !defined(IMGUI_DISABLE_WIN32_FUNCTIONS) && !defined(IMGUI_DISABLE_WIN32_DEFAULT_IME_FUNCTIONS)
+#if defined(_WIN32) && !defined(IMGUI_DISABLE_WIN32_FUNCTIONS) && !defined(IMGUI_DISABLE_WIN32_DEFAULT_IME_FUNCTIONS)
```
Comment (ptxmac) [inline imgui/imgui.cpp:10693]: This is where the old version was automatically disabled because github ci compiles with Microsoft's compiler.

File: imgui/imgui.h [hunk 2]
```diff
@@ -936,7 +946,7 @@ enum ImGuiInputTextFlags_
     ImGuiInputTextFlags_AllowTabInput       = 1 << 10,  // Pressing TAB input a '\t' character into the text field
     ImGuiInputTextFlags_CtrlEnterForNewLine = 1 << 11,  // In multi-line mode, unfocus with Enter, add new line with Ctrl+Enter (default is opposite: unfocus with Ctrl+Enter, add line with Enter).
     ImGuiInputTextFlags_NoHorizontalScroll  = 1 << 12,  // Disable following the cursor horizontally
-    ImGuiInputTextFlags_AlwaysInsertMode    = 1 << 13,  // Insert mode
```
Comment (dertseha) [inline imgui/imgui.h:939]: Please add the corresponding Go constant `InputTextFlagsAlwaysOverwrite`.
In order to avoid a major versin bump, define the existing `InputTextFlagsAlwaysInsertMode` to be equal the new one, and mark it as "Deprecated", with hint of becoming removed in next major version.

File: imgui/imgui.h [hunk 3]
```diff
@@ -2312,21 +2282,27 @@ struct ImDrawListSplitter
     IMGUI_API void              SetCurrentChannel(ImDrawList* draw_list, int channel_idx);
 };
 
-enum ImDrawCornerFlags_
+// Flags for ImDrawList functions
+// (Legacy: bit 0 must always correspond to ImDrawFlags_Closed to be backward compatible with old API using a bool. Bits 1..3 must be unused)
+enum ImDrawFlags_
```
Comment (dertseha) [inline imgui/imgui.h:2287]: This is an incompatible change not reflected in the wrapper.
Please figure out a way to make a compatible extension in the wrapper - I'd like to avoid bumping major versions for any minute change in the library.
Please cross-check also all affected function signatures.

File: imgui/misc/freetype/imgui_freetype.h [hunk 4]
```diff
@@ -26,7 +26,8 @@ enum ImGuiFreeTypeBuilderFlags
     ImGuiFreeTypeBuilderFlags_Bold          = 1 << 5,   // Styling: Should we artificially embolden the font?
     ImGuiFreeTypeBuilderFlags_Oblique       = 1 << 6,   // Styling: Should we slant the font, emulating italic style?
     ImGuiFreeTypeBuilderFlags_Monochrome    = 1 << 7,   // Disable anti-aliasing. Combine this with MonoHinting for best results!
-    ImGuiFreeTypeBuilderFlags_LoadColor     = 1 << 8    // Enable FreeType color-layered glyphs
+    ImGuiFreeTypeBuilderFlags_LoadColor     = 1 << 8,   // Enable FreeType color-layered glyphs
+    ImGuiFreeTypeBuilderFlags_Bitmap        = 1 << 9    // Enable FreeType bitmap glyphs
```
Comment (dertseha) [inline imgui/misc/freetype/imgui_freetype.h:30]: Please add this new constant in `FreeType.go`

**Review by dertseha — CHANGES_REQUESTED** [review]
Hello there and thank you for both your contribution and patience.
The changes in **Dear ImGui** include some incompatibilities that need to be addressed first. Please have a look at the comments and amend the PR with possible solutions that avoid a bump in major version of the wrapper.

File: imgui/imgui.h (hunk 3 above)
Comment (ptxmac) [reply imgui/imgui.h:2287]: Good catch. I guess all functions receiving the flag needs to convert from the old to the new.
Should we also add new functions that accept the new enum?

Comment (dertseha) [reply imgui/imgui.h:2287]: If it's possible to keep the same functions, yet with the new type, that would be great.

It should be by having the new flag type with the old constants being of that new type, in parallel to the new constants. The functions have the new flag type in their signature. Unless someone did some explicit typing in their code, this should work without a compilation issue.

In sum I'm OK with some compatibility redirects that work "automatically". I want to avoid creating an entire wrapper for compatibility in what is already a wrapper. 

So, something like:
```
type NewFlags int
type OldFlags int

const (
  NewFlagsConstant1 = NewFlags 1
  NewFlagsConstant2 = NewFlags 2
  ...
  // OldFlagsConstant1 does ...
  // Deprecated: Use NewFlagsConstant1. Will be removed in v5.
  OldFlagsConstant1 = NewFlagsConstant1 -- if applicable
)

func SomeFunc(..., flags NewFlags, ...) -- changing the signature from OldFlags
```

It's tricky to keep up with version compatibility and meaningful semantic versioning if the underlying library changes so often :/

File: Widgets.go [hunk 5]
```diff
@@ -566,8 +566,10 @@ const (
 	InputTextFlagsCtrlEnterForNewLine InputTextFlags = 1 << 11
 	// InputTextFlagsNoHorizontalScroll disables following the cursor horizontally.
 	InputTextFlagsNoHorizontalScroll InputTextFlags = 1 << 12
-	// InputTextFlagsAlwaysInsertMode sets insert mode.
+	// Deprecated: InputTextFlagsAlwaysInsertMode was renamed to InputTextFlagsAlwaysOverwriteMode in imgui v1.82 and will be removed in the next major version.
```
Comment (dertseha) [inline Widgets.go:569]:
```suggestion
	// InputTextFlagsAlwaysInsertMode was renamed to InputTextFlagsAlwaysOverwriteMode to reflect its actual behavior and will be removed in v5.
	// Deprecated: Use InputTextFlagsAlwaysOverwriteMode.
```

File: imgui/imgui.h [hunk 6]
```diff
@@ -2769,6 +2750,66 @@ struct ImGuiViewport
...
+    static inline void  TreeAdvanceToLabelPos()             { SetCursorPosX(GetCursorPosX() + GetTreeNodeToLabelSpacing()); }
+    // OBSOLETED in 1.71 (from June 2019)
+    static inline void  SetNextTreeNodeOpen(bool open, ImGuiCond cond = 0) { SetNextItemOpen(open, cond); }
+    // OBSOLETED in 1.70 (from May 2019)
+    static inline float GetContentRegionAvailWidth()        { return GetContentRegionAvail().x; }
+    // OBSOLETED in 1.69 (from Mar 2019)
+    static inline ImDrawList* GetOverlayDrawList()          { return GetForegroundDrawList(); }
+}
+
+// OBSOLETED in 1.82 (from Mars 2021): flags for AddRect(), AddRectFilled(), AddImageRounded(), PathRect()
```
Comment (ptxmac) [inline imgui/imgui.h:2795]: Just noticed that imgui also did a backwards-compatible change for the corner flags, so I'll try to replicate that

File: imgui/imgui.h (hunk 3 above)
Comment (ptxmac) [reply imgui/imgui.h:2287]: Did a attempt at this, it was easier than expected because imgui also tries to maintain backwards compatibility, so I could just copy the upstream definitions

File: go.mod [hunk 7]
```diff
@@ -1,5 +1,9 @@
 module github.com/inkyblackness/imgui-go/v4
 
-require github.com/stretchr/testify v1.3.0
+require (
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/stretchr/testify v1.7.0
```
Comment (ptxmac) [inline go.mod:5]: I also upgraded testify, let me know if that should be a different PR

Comment (ptxmac) [reply go.mod:5]: Never mind that, I forgot that newer testify versions require go1.13+

**Review by dertseha — APPROVED** [review]
great, thank you for the amendments!
